### ISSUE TITLE
Error when using custom statistics

### ISSUE THREAD

**Issue by piklprado** [lageIBUSP/Rsampling-shiny#16]
Using rhyzophora and custom statistics:

``` r
m1 = lm(n.roots ~ canopy.trunk , data=dataframe, subset=soil.instability=="high")
m2 = lm(n.roots ~ canopy.trunk , data=dataframe, subset=soil.instability=="medium")
coef(m1)[2]-coef(m2)[2]
```

An error is returned
`ERROR: unused argument (replace)`

Might be a problem with Rsampling and not Rsampling-shiny, but I could not find out. Any guess?

**Comment by andrechalom** [comment]
I'm getting several different problems when using rhyzophora with a custom function. The function `{return(pi)}` is throwing the following segfault:

``` R
 *** caught segfault *** 
address 0x4e0, cause 'memory not mapped'
Traceback:
 1: .Call("dplyr_mutate_impl", PACKAGE = "dplyr", df, dots)
 2: mutate_impl(.data, dots)
 3: mutate_.tbl_df(.data, .dots = lazyeval::lazy_dots(...))
 4: mutate_(.data, .dots = lazyeval::lazy_dots(...))
 5: mutate(., value = sample(value, size = n(), replace))
 6: function_list[[i]](value)
 7: freduce(value, `_function_list`)
 8: `_fseq`(`_lhs`)
 9: eval(expr, envir, enclos)
10: eval(quote(`_fseq`(`_lhs`)), env, env)
11: withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
12: cbind(rn = 1:nrow(dataframe), stratum = stratum, dataframe[cols]) %>%     as.data.frame() %>% gather("variable", "value", -rn, -stratum) %>%     group_by(stratum) %>% mutate(value = sample(value, size = n(),     replace)) %>% ungroup() %>% spread(variable, value) %>% select(-rn,     -stratum) %>% {    if (length(cols) < ncol(dataframe))         cbind(., dataframe[-cols])[colnames(dataframe)]    else .} %>% as.data.frame() 
13: (function (dataframe, cols = 1:ncol(dataframe), stratum = rep(1,     nrow(dataframe)), replace = FALSE) {    if (class(dataframe) != "data.frame")         stop("the 1st argument is not of class 'data.frame'")    cbind(rn = 1:nrow(dataframe), stratum = stratum, dataframe[cols]) %>%         as.data.frame() %>% gather("variable", "value", -rn,         -stratum) %>% group_by(stratum) %>% mutate(value = sample(value,         size = n(), replace)) %>% ungroup() %>% spread(variable,         value) %>% select(-rn, -stratum) %>% {        if (length(cols) < ncol(dataframe))             cbind(., dataframe[-cols])[colnames(dataframe)]        else .    } %>% as.data.frame()})(dataframe = list(soil.instability = c(1L, 1L, 1L, 1L, 1L, 1L, 1L, 1L, 1L, 1L, 1L, 1L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L), canopy.trunk = c(937.6087379973, 1957.0433922997, 1403.0969230769, 785.686875, 1431.3667666282, 1208.7815500686, 942.4617647059, 857.6845918367, 988.6761432507, 1219.6173261141, 365.1703703704, 279.9227839089, 1426.1850182615, 3675.3681331748, 1743.874829932, 721.9677514793, 501.7832142857, 2001.8971261974, 1036.8795221172, 1098.6411428572, 1171.200969697, 905.4734693878, 1244.31233125, 413.2590963139), root = c(19.782, 23.66775, 20.724, 6.7353, 15.7, 16.9717, 40.5531, 0.6990425, 18.3062, 15.6372, 6.0445, 8.080947, 6.7378905, 14.5225, 15.19132, 22.4667, 6.032725, 15.00135, 1.538757, 16.63101, 3.939444, 0.52124, 5.644464, 5.2332025), n.roots = c(117L, 152L, 106L, 91L, 161L, 125L, 131L, 24L, 172L, 107L, 53L, 86L, 77L, 108L, 102L, 93L, 68L, 61L, 36L, 81L, 33L, 19L, 55L, 36L)), cols = 1:4, stratum = c(1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1), replace = FALSE) 
14: do.call(f1, c(list(dataframe = dataframe), dots))
15: as.data.frame(my.df)
16: statistics(do.call(f1, c(list(dataframe = dataframe), dots)))
17: f() 
18: .rlply_worker(.n, .progress, eval.parent(substitute(function() .expr)))
19: rlply(ntrials, statistics(do.call(f1, c(list(dataframe = dataframe),     dots))), .progress = progress)
20: eval(expr, envir, enclos)
21: eval(lhs, parent, parent)
22: rlply(ntrials, statistics(do.call(f1, c(list(dataframe = dataframe),     dots))), .progress = progress) %>% {    if (simplify)         simplify2array(.)    else .}
23: Rsampling(type = type, dataframe = data(), statistics = statistic(),     cols = cols(), stratum = isolate(stratum()), ntrials = isolate(input$ntrials),     replace = isolate(input$replace))
24: .func()
(... shiny part of the stack ...)
```

It might be related to how the cols argument is being handled??
(EDIT: fixed! See below)

**Comment by andrechalom** [comment]
OK, running this function from the command line seems to fail as well, but with a different message.

``` R
> library(Rsampling)
> f <- function(dataframe)
+ {
+ m1 = lm(n.roots ~ canopy.trunk , data=dataframe, subset=soil.instability=="high")
+ m2 = lm(n.roots ~ canopy.trunk , data=dataframe, subset=soil.instability=="medium")
+ coef(m1)[2]-coef(m2)[2]
+ }
> Rsampling("norm", rhyzophora, f, 300, progress='none')
Error in lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) : 
  NA/NaN/Inf in 'y'
In addition: Warning messages:
1: attributes are not identical across measure variables; they will be dropped 
2: In model.response(mf, "numeric") : NAs introduced by coercion
```

With different random seeds, I also got the following message:

``` R
> Rsampling("norm", rhyzophora, f, 300, progress='none')
Error in `contrasts<-`(`*tmp*`, value = contr.funs[1 + isOF[nn]]) : 
  contrasts can be applied only to factors with 2 or more levels
```

(EDIT: fixed! This need explicit "cols" to work, see below)

**Comment by andrechalom** [comment]
Using rows as units in the command line leads to the error you have originally reported. However, the error is being triggered in the call to normal_rand??? But normal_rand DOES have a replace formal.

``` R
> Rsampling("rows_as_units", rhyzophora, f, 300, progress='none', stratum=rhyzophora$soil.instability)
Error in normal_rand(., function(x) x, cols = 1, stratum = stratum, replace) : 
  unused argument (replace)
```

**Comment by andrechalom** [comment]
The code for rows_as_units is

``` R
function (dataframe, stratum = rep(1, nrow(dataframe)), replace = FALSE) 
{
    if (class(dataframe) != "data.frame") 
        stop("the 1st argument is not of class 'data.frame'")
    cbind(ri = 1:nrow(dataframe), dataframe) %>% as.data.frame() %>% 
        normal_rand(function(x) x, cols = 1, stratum = stratum, 
            replace) %>% arrange(ri) %>% select(-ri) %>% as.data.frame()
}
```

Why is the first argument being passed as a function, instead of as a dataframe? I'm not very familiar with plyr syntax, but shouln't this be something like

``` R
    normal_rand(., cols=1, stratum=stratum, replace=replace) %>%
```

Also, notice the replace=replace.

**Comment by andrechalom** [comment]
OK, I'm not being able to get rows_as_units to work as I expect it to. Am I missing something?
(I'm using my edited version of rows_as_units, as the base function just throws errors)

``` R
> names(rhyzophora)
[1] "soil.instability" "canopy.trunk"     "root"             "n.roots"         
> f <- function(dataframe) mean(dataframe$root)
> f(rhyzophora)
[1] 12.7651
> # This produce the expected distribution:
> Rsampling("normal", cols=3, rhyzophora, f, 300, progress='none', replace=TRUE)->x
> summary(x)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  8.189  11.470  12.730  12.860  14.080  18.150 
# Now this should do THE SAME as the above, shouldn't it?
> Rsampling("rows", rhyzophora, f, 300, progress='none', replace=TRUE)->x
> summary(x)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  12.77   12.77   12.77   12.77   12.77   12.77 
```

EDIT: the code on the _noplyr_ branch works as expected:

``` R
> f <- function(dataframe) mean(dataframe$root)
>  Rsampling("normal", cols=3, rhyzophora, f, 300, progress='none', replace=TRUE)->x
> summary(x)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  6.727  11.470  12.790  12.750  14.020  17.330 
> Rsampling("rows", rhyzophora, f, 300, progress='none', replace=TRUE)->x
> summary(x)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  7.879  11.360  12.570  12.560  13.710  18.520 
```

**Comment by andrechalom** [comment]
Within columns and normal_rand lead to another problem: all the data types are converted to character if "cols" is not specified. This can cause some confusion.

``` R
> f <- function(dataframe) cat(str(dataframe))
> Rsampling("within_c", rhyzophora, f, 1, progress='none')->x
'data.frame':   24 obs. of  4 variables:
 $ soil.instability: chr  "medium" "high" "medium" "high" ...
 $ canopy.trunk    : chr  "1957.0433922997" "1426.1850182615" "1036.8795221172" "279.9227839089" ...
 $ root            : chr  "6.7378905" "1.538757" "8.080947" "22.4667" ...
 $ n.roots         : chr  "93" "19" "172" "33" ...
```

**Comment by andrechalom** [comment]
OK, I think we have three issues here. I`ll try to summarize the status so far:

### Interface issues
- segfault was being caused by a code that circumvented the spurious "stratum" column (fixed in Rsampling 0.0.0.2, so I'm removing the code in Rsampling-shiny. Fixed)
- We **need** to have a "cols" selector together with the custom f() to be able to run normal and whithin_cols on data.frames with more than one data type. (Fixed)

### Base package issues
- rows_as_units is throwing "ERROR: unused argument (replace)". MAYBE this can be fixed with

``` R
    normal_rand(., cols=1, stratum=stratum, replace=replace) %>%
```

**Comment by andrechalom** [comment]
The example above is working now for normal_rand. The problem within_rows seems to me to be a bug in the base package, so I'm closing this issue as an interface bug

![untitled](https://cloud.githubusercontent.com/assets/1994997/8487422/661da358-20e2-11e5-91cd-665935432c58.png)
